### PR TITLE
Update lastFlushedIndex in metastore in Journal

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

This PR is in preparation to allowing async flush #11488. 
- Updating `lastFlushedIndex` is moved to Journal. This is needed because with async/configurable flush interval RaftContext doesn't know when flush will be completed and cannot reliably update lastFlushedIndex.
- It still uses the same MetaStore as before, but is hidden under an interface which is provided to the Journal.

Temporarily, in this PR metastore is updated in `SegmentedJournal::flush`. This might change depending on the implementation of async flush. A followup PR will refactor the use of `lastFlushedIndex` in the journal to use it only on startup. The goal is not to pass it all around in segment and segment writers. 

## Related issues

related #11488

### CODE REVIEW

**Review by npepinpe — COMMENTED** [review]
:rocket: 

- :heart: Javadoc for the `JournalMetaStore`
- :heart: Not passing around the last written/flushed index everywhere anymore
- :x: Sorry if I missed it, but I didn't see a test that when the journal flushes the index is updated properly. I can also add it in the subsequent PR though.
- :question: I thought this PR would also remove passing the index for the segment writers, and only checking for corruption when reading/during the initial open of the journal for the writer. Are we doing this in a separate PR?

**Comment by deepthidevaki** [comment]
> * x Sorry if I missed it, but I didn't see a test that when the journal flushes the index is updated properly. I can also add it in the subsequent PR though.

There is a new test  `SegmentedJournal::shouldUpdateMetastoreAfterFlush`.

> * question I thought this PR would also remove passing the index for the segment writers, and only checking for corruption when reading/during the initial open of the journal for the writer. Are we doing this in a separate PR?

To minimize the scope, I will do it as a separate PR as mentioned in the PR description.

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/MetaStore.java [hunk 1]
```diff
@@ -199,6 +191,17 @@ public synchronized void storeLastFlushedIndex(final long index) {
     }
   }
 
+  @Override
+  public synchronized long loadLastFlushedIndex() {
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/system/MetaStore.java:195]: Yes. But metastore will be closed after the RaftLog.

**Comment by deepthidevaki** [comment]
@npepinpe I will add followup PR for removing lastFlushedIndex for Segment Writer. So we can merge this PR now. Could you approve?

**Review by npepinpe — APPROVED** [review]
:+1:

**Pull request merged**